File: 6/construct_str_from_btree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def tree2str(self, t: TreeNode) -> str:
        if not t:
            return ""
        def print_node(t: TreeNode):
            if not t.left and not t.right:
                return "{}".format(t.val)
            elif t.left and not t.right:
                return "{}({})".format(t.val, print_node(t.left))
            elif t.right and not t.left:
                return "{}()({})".format(t.val, print_node(t.right))
            else:
                return "{}({})({})".format(t.val, print_node(t.left), print_node(t.right))
        return print_node(t)

File: 6/construct_str_from_btree.go
package main

import (
	"fmt"
)

type Node struct {
	val   int
	left  *Node
	right *Node
}

func tree2str(n *Node) string {
	if n == nil {
		return ""
	}

	if n.left == nil && n.right == nil {
		return fmt.Sprintf("%d", n.val)
	} else if n.left == nil {
		return fmt.Sprintf("%d()(%s)", n.val, tree2str(n.right))

	} else if n.right == nil {
		return fmt.Sprintf("%d(%s)", n.val, tree2str(n.left))
	} else {
		return fmt.Sprintf("%d(%s)(%s)", n.val, tree2str(n.left), tree2str(n.right))
	}
}

func main() {
	n21 := Node{4, nil, nil}
	n11 := Node{2, &n21, nil}
	n12 := Node{3, nil, nil}
	root := Node{1, &n11, &n12}
	fmt.Println(tree2str(&root))
}
